ncm-spma/src/main/perl/spma/apt.pm

```diff
@@ -0,0 +1,433 @@
+
+    while (my $name = shift(@packages)) {
+        my $name_escaped = escape($name);
+        if (exists($packagetree->{$name_escaped})) {
+            my $versions = $self->get_package_version_arch($name, $packagetree->{$name_escaped});
+            push(@results, @$versions);
+        }
+        else {
+            $self->error('Could not find package ', $name, ' in tree');
+        }
+    }
+
+    return \@results;
+}
+
+
+# Return a set of desired packages.
+sub get_desired_pkgs
+{
+    my ($self, $pkgs) = @_;
+    $self->debug(5, 'Entered get_desired_pkgs()');
+
+    my $packages = Set::Scalar->new();
+
+    while (my $name_escaped = each(%$pkgs)) {
+        my $name = unescape($name_escaped);
+        if (!$name) {
+            $self->error("Invalid package name: ", unescape($name));
+            return 0;
+        }
+        $packages->insert($name);
+    }
+    return $packages;
+}
+
+
+# Update package metadata from upstream repositories
+sub resynchronize_package_index
+{
+    my $self = shift;
+    $self->debug(5, 'Entered resynchronize_package_index()');
+
+    my $cmd = CAF::Process->new($CMD_APT_UPDATE, keeps_state => 1, log => $self);
+    $cmd->execute() or return 0;
+
+    return 1;
+}
+
+
+# Upgrade existing packages
+sub upgrade_packages
+{
+    my ($self) = @_;
+    $self->debug(5, 'Entered upgrade_packages()');
+
+    my $cmd = CAF::Process->new($CMD_APT_UPGRADE, keeps_state => 1, log => $self);
+    $cmd->execute() or return 0;
+
+    return 1;
+}
+
+
+# Install packages
+sub install_packages
+{
+    my ($self, $packages) = @_;
+    $self->debug(5, 'Entered install_packages()');
+
+    my $cmd = CAF::Process->new([@$CMD_APT_INSTALL, @$packages], keeps_state => 1, log => $self);
+    $cmd->execute() or return 0;
+
+    return 1;
+}
+
+
+# Mark packages as automatically installed
+# this signals to apt that they are no longer required and may be cleaned up by autoremove
+sub mark_packages_auto
+{
+    my ($self, $packages) = @_;
+    $self->debug(5, 'Entered mark_packages_auto()');
+
+    my $cmd = CAF::Process->new([@$CMD_APT_MARK, 'auto', @$packages], keeps_state => 1, log => $self);
+    $cmd->execute() or return 0;
+
+    return 1;
+}
+
+
+# Remove automatically installed packages
+sub autoremove_packages
+{
+    my ($self, $packages) = @_;
+    $self->debug(5, 'Entered autoremove_packages()');
+
+    my $cmd = CAF::Process->new([@$CMD_APT_AUTOREMOVE, @$packages], keeps_state => 1, log => $self);
+    $cmd->execute() or return 0;
+
+    return 1;
+}
+
+
+sub Configure
+{
+    my ($self, $config) = @_;
+    $self->debug(5, 'Entered Configure()');
+
+    # Get configuration trees
+    my $tree_repos = $config->getElement($TREE_REPOS)->getTree();
+    my $tree_pkgs = $config->getElement($TREE_PKGS)->getTree();
+    my $tree_component = $config->getElement($TREE_COMPONENT)->getTree();
+    # Convert these crappily-defined fields into real Perl booleans.
+    $tree_component->{run} = $tree_component->{run} eq 'yes';
```

Comment:
If you intend to start converting these fields in other code to be either booleans or text, you should start that effort here by accepting both.